server/container/guice/guice-common/src/main/java/org/apache/james/GuiceJamesServer.java

```diff
@@ -78,6 +78,8 @@ public GuiceJamesServer overrideWith(List<Module> overrides) {
     public void start() throws Exception {
         Injector injector = Guice.createInjector(module);
         preDestroy = injector.getInstance(Key.get(new TypeLiteral<Stager<PreDestroy>>() {}));
+        injector.getInstance(StartUpChecksPerformer.class)
```

Comment:
what will happen in the case of a missing injection?

Reply:
do you mean: what happen in the case of no startup check injected?

Reply:
No, I mean something else.
You are using the main guice injector. When you will instantiate classes, if one of these classes has a dependency on something that fails to initialize, it will fail before running the checks.
I understand that you want to use guice to have an extensible list of checks and it's not a bad design.
However, you probably don't want to use the main injector. You could very well have a specific set of modules for these startupchecks. That way, you would not be able to depend on a server component.
By the way, the checks should depend on as few things as possible to ensure it actually checks thinks and not break itself.

Reply:
IMO having a separate injector represent a significant work and will add significant complexity (having to sets of module, explicitly specifying the modules needed for the tests, etc..)

I agree we need to pool as few things as possible in these checks, but guici init is lazy, and we can achieve the very same thing (partial minimal application load for checks) with a single injector. We should just be carefull when declaring the dependencies of a check - to not pull planet earth in.

Reply:
I'm looking for the cassandra migration checks, to do that I need the cassandra DAO, then I have to bind the cassandra executor, bind the DAO, and installing some other guice modules are pretty verbose and require much effort to do it with another injector.

But your argument makes sense, what do you think @chibenwa

Reply:
I think having a separate injector, with a subset of the main injector makes little sens:

 - These components will be initialized 2 times (eg cassandra session for checking version) resulting in : 
    - Harder lifecycle management (you need to close the Cassandra session you used to check the version)
    - Higher starting latencies (you initialize 2 time a Cassandra session and it takes time)
    - More code complexity - you need to maintain 2 sets of modules instead of one.

I do believe that the proposition from Matthieu adds little benefits: again injector are **lazy** and will init only the dependencies you need given the objects you require. Here **only what is strictly required for checks is pulled** and then is reused afterward upon application startup.

The only benefit I can see about making separated bindings  is *sensibilize DEVs about the amount of dependencies they pull*. Given the drawbacks listed above, my opinion is clear: let's not add that sparate injector now.

Reply:
The PR is closed and you made a choice that I'm not willing to override.
But I think answering to some arguments would still be useful, so here it is.

> These components will be initialized 2 times 

False. see Injector.createChildInjector.

> I'm looking for the cassandra migration checks, to do that I need the cassandra DAO, then I have to bind the cassandra executor, bind the DAO, and installing some other guice modules are pretty verbose and requires much effort to do it with another injector.

I don't get it: Guice is made of modules. we just have to select some modules to promote for first phase injector. If it's hard, we have problems in the way we use Guice.

> The only benefit I can see about making separated bindings is sensibilize DEVs about the amount of dependencies they pull

In my experience, we can expect errors here and the list of dependencies to be a lot bigger than what we want (and so startup checks will be slower than we it could be). Just have a look at maven: every time I look at the dependencies of a module, there are things we can drop.

Reply:
> see Injector.createChildInjector.

True. Thanks for bringing that in.

Turns out 'collection injections' like setMultibinding can not be shared between these stages. Hence difficulties arise for instance if you need some Cassandra modules for instance in your startup checks.

>  we just have to select some modules to promote for first phase injector.

Agree. However we'd have to bing complexity in our for now quite dumb GuiceJamesServer.

> In my experience, we can expect errors here and the list of dependencies to be a lot bigger than what we want

Agree, but I suspect people will just declare more "first step guice modules" to overcome the bureaucracy we putted in place to prevent that. Again, I know the issue will arise, and I don't know what the solution is.